### DISCUSSION TITLE
Why do we need a "completion engine", or, what is a "completion engine"?

### DISCUSSION THREAD

**Discussion by debdutdeb** [neovim/neovim#26639] [General]
Specifically talking about lsp, neovim already has lsp support. Completion also technically works with c-x c-o.

![image](https://github.com/neovim/neovim/assets/76006232/61e314be-a478-459a-a501-aff5e20af58c)

Here above for example I don't have any completion engine plugged in, like nvim_cmp or coq. And hitting c-x c-o works as one would expect.

Then why do we need an engine? What is the purpose of an engine?

Is it to just periodically call completion func in the background? I don't seem to find any documentation on this.

Sorry if it's too apparent. I'm trying to wrap my head around this :)

**Comment by clason** [comment]
Details vary, of course, but in general a "completion engine" 
1. enables **auto**completion (e.g., based on trigger characters)
2. supports several different (configurable) sources and merging results from them.
